flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/SchedulerBase.java

```diff
@@ -277,9 +282,39 @@ private void tryRestoreExecutionGraphFromSavepoint(ExecutionGraph executionGraph
 		}
 	}
 
-	protected void resetForNewExecutionIfInTerminalState(final Collection<ExecutionVertexID> verticesToDeploy) {
-		verticesToDeploy.forEach(executionVertexId -> getExecutionVertex(executionVertexId)
-			.resetForNewExecutionIfInTerminalState());
+	protected void resetForNewExecutions(final Collection<ExecutionVertexID> vertices) {
+		vertices.forEach(executionVertexId -> getExecutionVertex(executionVertexId)
+			.resetForNewExecution());
+	}
+
+	protected void restoreState(final Set<ExecutionVertexID> vertices) throws Exception {
+		// if there is checkpointed state, reload it into the executions
+		if (executionGraph.getCheckpointCoordinator() != null) {
+			// abort pending checkpoints to
+			// i) enable new checkpoint triggering without waiting for last checkpoint expired.
+			// ii) ensure the EXACTLY_ONCE semantics if needed.
+			executionGraph.getCheckpointCoordinator().abortPendingCheckpoints(
+				new CheckpointException(CheckpointFailureReason.JOB_FAILOVER_REGION));
+
+			final Map<JobVertexID, ExecutionJobVertex> involvedExecutionJobVertices =
+				getInvolvedExecutionJobVertices(vertices);
+
+			executionGraph.getCheckpointCoordinator().restoreLatestCheckpointedState(
```

Comment:
It seems that `restoreLatestCheckpointedState` does not require the `JobVertexID` at all. Maybe we can introduce a overload method?

Reply:
You are right. I made the change in 1d74f9981eca26f03d5daf610cb96d2e5174cb2b. All usages are also refactored except for those for the legacy scheduler.